Update Object3D.MatrixWorld to always be up to date

gero3 opened the pull request:
I dislike this one more then the previous one #6448 becuase now we have to maintain how to calculate matrixWorld in 2 different places:
- The get of MatrixWorld 
- The function UpdateMatrixWorld

gero3 committed:
add getters and caches voor matrixWOlrd and Matrix

mrdoob commented:
> becuase now we have to maintain how to calculate matrixWorld in 2different places:

I think we can address that.

@WestLangley how do you feel about this approach?

WestLangley commented:
> @WestLangley how do you feel about this approach?

Not sure, yet. I need some time to think about this version vs. #6448.

mrdoob commented:
With this approach, `object.matrix` and `object.matrixWorld` are always updated, so no need to tell people to update them. They also become immutable, just like `position`, `rotation`, ... which is tricky but I think better in the long term.

WestLangley commented:
OK. Now I understand the functionality that is important to you.

I still need to study it, but would it mean the following dreadful patteren would no longer be supported?

	this.matrix = object.matrixWorld;
	this.matrixAutoUpdate = false;

Also, will users still be able to set the `matrix` directly?

mrdoob commented:
No, `matrix` and `matrixWorld` will become immutable. That pattern would no longer work.

But, we can continue having `matrixAutoUpdate` and (advanced) users could do `this._matrix = object.matrixWorld`.

WestLangley commented:
I just do not see how this can work...

We should only recompute `matrix` or `matrixWorld` if it is required to do so -- not every time the matrix is accessed.

Also, we have to be able to support this pattern

    groupt.add( object );
    object.add( child );

    group.postion.x += 1;
    vector.setFromMatrixPosition( child.matrixWorld );

The above pattern is the use-case that has never been properly supported. A proposed solution has been to call `updateMatrixWorld()` prior to calling `setFromMatrixPosition()`, but that solution requires modifying `updateMatrixWorld()` to bubble up the scene graph.

mrdoob commented:
Something that we could consider is using [Object.observe()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/observe#Browser_compatibility)

Even if it's only supported in Chrome... It could be a way to encourage other browsers to support it.

So the code will just look like this:

var position = new THREE.Vector3();
var position = new THREE.Vector3();

Object.observe( position, setMatrixDirty );

mrdoob commented:
Another option would be something like this...

```javascript
var position = new THREE.Vector3();
var rotation = new THREE.Euler();
var scale = new THREE.Vector3();

var matrix = new THREE.Matrix4();

var lastState = position.toString() + rotation.toString() + scale.toString();

Object.defineProperties( this, {
	matrix: {
		enumerable: true,
		get: function () {
			var state = position.toString() + rotation.toString() + scale.toString();
			if ( state !== lastState ) {
				matrix.compose( position, quaternion, scale );
				lastState = state;
			}
			return matrix;
		}
	},
} );
```

mrdoob commented:
This last option may have potential. The question is whether doing

```javascript
[].concat( position.toArray(), quaternion.toArray(), scale.toArray() ).join( ',' )
```

is faster than doing

```javascript
matrix.compose( position, quaternion, scale );
```

mrdoob commented:
I wonder whether `[].concat( position.toArray(), quaternion.toArray(), scale.toArray() ).join( ',' )` is worth it. Otherwise the code could be simpler:

```javascript
var position = new THREE.Vector3();
var quaternion = new THREE.Quaternion();
var scale = new THREE.Vector3();

var matrix = new THREE.Matrix4();
var matrixWorld = new THREE.Matrix4();

var lastState = matrix.toArray().join(',');

function updateMatrix() {
	matrix.compose( position, quaternion, scale );
	var state = matrix.toArray().join(',');
	if ( state !== lastState ) {
		this.matrixWorldNeedsUpdate = true;
		this.traverse( function ( child ) {
			child.matrixWorldNeedsUpdate = true;
		} );
		lastState = state;
	}
}

Object.defineProperties( this, {
	matrix: {
		enumerable: true,
		get: function () {
			updateMatrix();
			return matrix;
		}
	},
	matrixWorld: {
		enumerable: true,
		get: function () {
			if ( this.matrixAutoUpdate === true ) updateMatrix();
			if ( this.matrixWorldNeedsUpdate === true ) {
				if ( this.parent === undefined ) {
					matrixWorld.copy( matrix );
				} else {
					matrixWorld.multiplyMatrices( this.parent.matrixWorld , matrix );
				}
				this.matrixWorldNeedsUpdate = false;
				return matrixWorld;
			}
		}
	}
} );
```

gero3 committed:
keep track of states

gero3 committed:
make parents update before other matrixworlds

gero3 commented:
I've updated this branch to include laststate.

Difference:
 - It saved the laststate as a array of numbers because this should be faster then tostring and a lot faster then compose.
- `matrixWorld` is internally calls `updateMatrixWorld` with no recursivity and with updated parents.
- I've kept the updateMatrixWorld force parameter but added 
   - recursive which is default true
   - Parentchanged which is default true( this is nessecary to not update the parents when updating the children also)

WestLangley commented:
Thanks, @gero3, for working on this. I have not yet had the opportunity to carefully study your code, but I would like to make some comments.

`updateMatrixWorld()` is a method originally written for the renderer to apply to the scene:

`scene.updateMatrixWorld()`.

It is an efficient, depth-first, traversal of the scene graph.

We later advised users to call `object.updateMatrixWorld()` to update the world matrix of an object whenever the world matrix was used in a calculation.

Unfortunately, when used in that way, the method

1. did not necessarily produce the correct result, because parents were not updated first, and 
2. it unnecessarily update the world matrices of children.

We need to do (1) always, and and not do (2) by default.

I know this changes the behavior, but I have never seen a use case where the traversal of children was expected, except when the method was applied to the scene.

If you want to create multiple methods, that would be OK.

My request would be for `object.updateMatrixWorld()` to do just what its name says it does -- (correctly) update the world matrix of the object (only).

mrdoob commented:
@WestLangley I'm basically proposing removing `.updateMatrix()` and `.updateMatrixWorld()` altogether :)

gero3 commented:
Basically I added @mrdoob matrix & matrixworld but they call updateMatrix and updateMatrixWorld.
The reason I don't want to drop updateMatrixWorld is because getters are still a lot slower then calling function. We can use cached version of the matrices in the renderer after calling updateMatrixWorld. And users can always be sure that matrixWorld is up to date to use it in their application.

Also I checked out the performance of my version here compared with r71 updateMatrixWorld. 
[My version](https://jsfiddle.net/mpLw9n8b/1/)
[r71](https://jsfiddle.net/dkzebdc4/1/)

Watch console to see the timings.

gero3 committed:
don't create a new function every time updateMatrix is called.

gero3 committed:
no need to traverse the objects

mrdoob commented:
> The reason I don't want to drop updateMatrixWorld is because getters are still a lot slower then calling a function.

Sounds good yes!

mrdoob commented on src/core/Object3D.js:
```diff
@@ -534,44 +547,104 @@ THREE.Object3D.prototype = {
+		return function(){		
+			if 	(checkLastStateChanged(this)){
+				this.matrixCached.compose( this.position, this.quaternion, this.scale );
+				this.matrixWorldNeedsUpdate = true
```
How about also doing

```javascript
this.traverse( function ( child ) {
child.matrixWorldNeedsUpdate = true;
} );
```

I think this is what you're doing with `parentChanged` in `updateMatrixWorld` but seems cleaner?

gero3 replied:
The problem with that is that when calling `updateMatrixworld` on a child object and the position of the parent is changed (without calling `updateMatrix`) then it wouldn't update the `matrixWorld` automatically for the child because it wouldn't see the that something has changed.

mrdoob replied:
> The problem with that is that when calling `updateMatrixworld` on a child object and the position of the parent is changed (without calling `updateMatrix`)

But wouldn't the `updateWorldMatrix` call `updateMatrix` and update depending on `checkLastStateChanged`?

gero3 commented:
The reason I don't want 2 functions that update matrixWorld is because of code like: 
 - https://github.com/mrdoob/three.js/blob/master/src/extras/audio/AudioListener.js#L18-L49
 - https://github.com/mrdoob/three.js/blob/master/src/extras/core/Gyroscope.js#L14-L65
 - https://github.com/mrdoob/three.js/blob/master/src/objects/SkinnedMesh.js#L137-L155

gero3 committed:
no need to force anymore after update of matrixWorldCached

gero3 committed:
add traversal to make sure everything works

mrdoob commented:
> The reason I don't want 2 functions that update matrixWorld is because of code like: 

Ah, right. Good catch yes.

gero3 committed:
cleanup

gero3 committed:
use object instead of an array to keep the laststate

gero3 commented:
@WestLangley @mrdoob 

It seems to work well now compared to r71.
On 50000 objects in a scene, it is about 10 to 25% slower for moving all objects before recalculating and about 2 to 3 times faster when nothing is changed. 

PS: Updated test cases 
[My version](http://gero3.github.io/test_updateworldmatrix/r72MatrixWorld/index.html)
[r71](http://gero3.github.io/test_updateworldmatrix/r71/index.html)

gero3 committed:
mrdoobapproves

WestLangley commented:
> I know this changes the behavior, but I have never seen a use case where the traversal of children was expected, except when the method was applied to the scene.

What was I thinking? A method I wrote expects it: `Box3.setFromObject()`.  :/

1. In spite of that, I still think updating children should not be the default -- as it will not typically be required by a user.

2. Also, as an aside, would it make sense to have setters for `position`, `quaternion/rotation`, and `scale` that sets a dirty flag or version when changed -- instead of `setLastState()`?

3. And how about renaming `matrixCached / matrixWorldCached` to `_matrix / _matrixWorld` -- following the underscore pattern in `Quaternion`?

gero3 commented:
> I know this changes the behavior, but I have never seen a use case where the traversal of children was expected, except when the method was applied to the scene.
> What was I thinking? A method I wrote expects it: Box3.setFromObject(). :/
> 
> In spite of that, I still think updating children should not be the default -- as it will not typically be required by a user.

I agree. I'll change it but I wanted to keep this version first as close as possible to the original one.

> Also, as an aside, would it make sense to have setters for position, quaternion/rotation, and scale that sets a dirty flag or version when changed -- instead of setLastState()?

The problem with that is that most of the time we change the x,y,z,w of quaternion, position and scale. Which means we need  create setters for them too. which would complicate things quite a bit. 

> And how about renaming matrixCached / matrixWorldCached to _matrix / _matrixWorld -- following the underscore pattern in Quaternion?

I actually was thinking that matrixCached/matrixWorldCached were allowed to be used in userland and  _matrix / _matrixWorld would suggest that they only are to be used internally. But I'll change them for now. It is easy enough to change later.

WestLangley commented:
> _matrix / _matrixWorld would suggest that they only are to be used internally.

Given what you propose here, I belive that is true.

As I mentioned previously, this PR will cause breakage. The following pattern will no longer work (I think):

    this.matrix = object.matrixWorld;
    this.matrixAutoUpdate = false;

gero3 commented:
@WestLangley That can be fixed too as following:

	get matrix () {
		if ( this.matrixAutoUpdate === true ) {
			return this.updateMatrix();
		} else {
			return this._matrix;
		}
	},
	set matrix (value) {
		this._matrix.copy(value);
	},

What do you think @mrdoob??

gero3 committed:
default to non recursive

gero3 committed:
switch to underscore instead of cached

gero3 committed:
matrixAutoUpdate should be more useful now.

gero3 committed:
update all classes to make it work

gero3 commented:
I've updated this branch to fix all raised issues by @WestLangley. Even 

    this.matrix = object.matrixWorld;
    this.matrixAutoUpdate = false;

can be used.

The only thing I didn't change is setting setters.

WestLangley commented on src/extras/helpers/VertexNormalsHelper.js:
```diff
@@ -61,7 +61,7 @@ THREE.VertexNormalsHelper.prototype.update = ( function () {
 		var keys = [ 'a', 'b', 'c' ];
 
-		this.object.updateMatrixWorld( true );
+		this.object.updateMatrixWorld();
```
If `object.matrixWorld` is lazily updated whenever it is accessed, why do we need to call `updateMatrixWorld` -- ever?

gero3 replied:
We don't have to actually only when we want to precalculate the matrices

WestLangley replied:
Is your intention to remove all calls to `updateMatrixWorld()`, then -- with the exception of the renderer, perhaps?

gero3 replied:
I would.

WestLangley commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1526,7 +1526,7 @@ THREE.WebGLRenderer = function ( parameters ) {
 		// update scene graph
 
-		if ( scene.autoUpdate === true ) scene.updateMatrixWorld();
+		if ( scene.autoUpdate === true ) scene.updateMatrixWorld( );
```
Hmmm. I would have expected you would only call `updateMatrixWorld()` here if you wanted to traverse the entire tree and update all transforms for efficiency or performance reasons... If so, you need to force children to be updated.

If you only want to update the scene's matrix, then we are in trouble, because there is breakage if a user transforms the scene -- for example, the helpers will not work.

gero3 replied:
Indeed I missed this because it still works without updating everything first.

WestLangley commented:
@gero3 This is looking good. We are going to have to debate some details to finalize it. I need to do this step-by-step, so please be patient with me. : - )

mrdoob commented:
> As I mentioned previously, this PR will cause breakage. The following pattern will no longer work (I think):

>    this.matrix = object.matrixWorld;
>    this.matrixAutoUpdate = false;

Don't you think this is a pattern we should discourage? I know that I was the one that started it anyway. I guess it's ok if we break that, it will force us to figure out a nicer/cleaner way.

WestLangley commented:
> this.matrix = object.matrixWorld;
this.matrixAutoUpdate = false;
>> Don't you think this is a pattern we should discourage? I know that I was the one that started it anyway.
I guess it's ok if we break that, it will force us to figure out a nicer/cleaner way.

@gero3 has responded to my concerns about that pattern and provided a solution with this PR.

Be that as it may, I agree. I am not a fan of that pattern.

mrdoob commented on src/core/Object3D.js:
```diff
@@ -72,6 +76,18 @@ THREE.Object3D = function () {
 	this.userData = {};
+	
+	//contains the last state of position, quaternion & scale
+	this._matrixState = { positionx: 0, 
```
While we're at it... How about...

```javascript
this._matrixState = {
    position: position.clone(),
    quaternion: quaternion.clone(),
    scale: scale.clone()
};

// and ...

function checkLastStateChanged( position, quaternion, scale, matrixState ) {
			
	return matrixState.position.equals( position ) &&
		matrixState.quaternion.equals( quaternion ) &&
		matrixState.scale.equals( scale );

};

function setLastState( position, quaternion, scale, matrixState ) {

	matrixState.position.copy( position );			
	matrixState.quaternion.copy( quaternion );
	matrixState.scale.copy( scale );
				
};
```

gero3 replied:
I'll look into it but I think performance will degrade.

gero3 commented:
> this.matrix = object.matrixWorld;
> this.matrixAutoUpdate = false;
> 
> Don't you think this is a pattern we should discourage? I know that I was the one that started it anyway.
> I guess it's ok if we break that, it will force us to figure out a nicer/cleaner way.
> @gero3 has responded to my concerns about that pattern and provided a solution with this PR.
> 
> Be that as it may, I agree. I am not a fan of that frequently-used pattern.

I wouldn't break it for now. I now always update matrixworld when matrixAutoUpdate is false which is logical to me.

mrdoob commented:
I just gave a go at this myself. I think this is a bit too complicated at the moment.

I would suggest to just implement `object.getUpdatedMatrixWorld()` by now and implement in the relevant cases.

Sorry for the back and forth guys!

WestLangley commented:
@mrdoob I think @gero3's implementation is working... although I am still in the process of arguing for some small changes.

However, if you do not want to implement this PR as written, then what we should do for now is:

1. Modify `Object3D.updateMatrixWorld()` so it traverses parents by default and omits children by default when no args are set.
2. Have the renderer continue to call `scene.updateMatrixWorld( ..., ..., ... )`, forcing children to be traversed, instead of parents.
3. Continue to advise users the of use cases in which they must call `updateMatrixWorld()` between render loops.

And if possible, get rid of the `force` argument if possible.

gero3 commented:
> And if possible, get rid of the force argument to updateMatrixWorld().

it is possible. Even this change is possible in this pull request

mrdoob commented:
Ok. I'll wait until you guys think is ready 😇

gero3 commented:
> I just gave a go at this myself. I think this is a bit too much change at once.
> 
> I would suggest to just implement object.getUpdatedMatrixWorld() by now and replace the relevant this.updateMatrixWorld( true )s. (Step by step)
> 
> Sorry for the back and forth guys!

It isn't that big of a change actually. Because 95 percent of the users of three.js won't even notice the change.

And if we remove force, then it would simplify the code also.

mrdoob commented:
Yep yep! Curious to see how the code logic is without `force`. At least now I have a better understanding of all this.

WestLangley commented:
OK, so it  looks like we are proceeding... 

@gero3 Would you mind fixing the issues I pointed out (if you agree), and and also resolving the conflicts. ( Unfortunately, I my git skills must be lacking, because I am having trouble forking your topic branch and rebuilding without errors. ;/ )

gero3 committed:
remove force and fix updateMatrixWorld

gero3 committed:
someminor fixes

gero3 committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into MatrixWorld

Conflicts:
	src/extras/audio/Audio.js
	src/extras/audio/AudioListener.js
	src/extras/core/Gyroscope.js

gero3 commented:
I removed force & updatematrixworld on the places it wasn't absolutely nessecary.
 
I also update the branch to the latest dev.

mrdoob commented:
Did you get to try the `this._position.equals( this.position ) && this._quaternion.equals( this.quaternion ) && this._scale.equals( this.scale )` idea?

WestLangley commented:
> @WestLangley I'm basically proposing removing .updateMatrix() and .updateMatrixWorld() altogether :)

In fact, now `updateMatrixWorld()` can be made private and renamed `_updateMatrixWorld()`. We can revisit that later if you want.

`updateMatrix()` is still required to be public for this pattern:

    mesh.updateMatrix();
    mesh.matrixAutoUpdate = false;

Maybe we could make it private and have a `Mesh.static` flag that a user sets, instead. It would be a setter that calls the above code.

We should investigate the `matrixWorldNeedsUpdate` flag and its usage. It should be private, I think.


> Were there any other issues to needed to be handled?? @WestLangley

This is looking pretty good. There are a lot of calls to `updateMatrix()` and `updateMatrixWorld()` in the code and examples than can be removed.

For `updateMatrixWorld( recursive, parentChanged )`, I'd rename `parentChanged` to `updateParents`. In fact, i'd consoder renaming `recursive` to `updateChildren` for absolute clarity. Remember, users should not have to call this.

gero3 commented:
> Did you get to try the this._position.equals( this.position ) && this._quaternion.equals( this.quaternion ) && this._scale.equals( this.scale ) idea?

not yet.

> This is looking pretty good. There are a lot of calls to updateMatrix() and updateMatrixWorld() in the code and examples than can be removed.
> 
> For updateMatrixWorld( recursive, parentChanged ), I'd rename parentChanged to updateParents. In fact, i'd consider renaming recursive to updateChildren for absolute clarity. Remember, users should not have to call this.

I'll update this. 

> We should investigate the matrixWorldNeedsUpdate flag and its usage. It should be private, I think.

I agree.

gero3 committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into MatrixWorld

Conflicts:
	examples/js/Mirror.js
	examples/js/WaterShader.js

gero3 committed:
rename recursive & parentChanged to updateChildren & updateParents

gero3 committed:
make matrixWorldNeedsUpdate private

gero3 committed:
use _position & _quaternion & _scale to test matrix changed

gero3 commented:
I've updated recursive and parentChanged to updateChildren and updateParents and made MatrixWorldNeedsUpdate into _MatrixWorldNeedsUpdate to make it private. Also i added _position & _quaternion and _scale as it is as fast as the other version. This also cleans up some functions inside updateMatrix.

I'm now looking to update updateMatrixWorld to _updateMatrixWorld.

gero3 committed:
more cleanup of updateMatrixWorld

gero3 commented:
Updating updateMatrixWorld to _updateMatrixWorld is harder then I expected.
Mostly becuase of code like this:

 - https://github.com/gero3/three.js/blob/MatrixWorld/src/objects/SkinnedMesh.js#L126-L138
 - https://github.com/gero3/three.js/blob/MatrixWorld/src/extras/audio/Audio.js#L130-L132
 - https://github.com/gero3/three.js/blob/MatrixWorld/src/extras/audio/AudioListener.js#L30-L38

They update other properties that are only reliant on matrix world. 
I could make getters for these properties. What do you think  @mrdoob @WestLangley

mrdoob commented:
I think we can keep `updateMatrixWorld` by now?

mrdoob commented on src/core/Object3D.js:
```diff
@@ -538,44 +560,95 @@ THREE.Object3D.prototype = {
+		} else {
+		
+			// first update parents before check matrixWorldNeedsUpdate
+			if ( updateParents !== false ) {
```
Could we make this check `updateParents === true`? This will make `updateMatrixWorld()` a bit more intuitive. (I'm aware that most of the calls will become `.updateMatrixWorld( true, true )` and `.updateMatrixWorld( false, true )`.

WestLangley replied:
We want `updateMatrixWorld()` with no args to update parents and not children.

gero3 replied:
I agree with @WestLangley on this one.

WestLangley commented:
> I could make getters for these properties. What do you think @mrdoob @WestLangley

We can revist it a later time.

mrdoob commented:
Ok, how about this?

```javascript
updateMatrixWorld: function () {

	if ( this.matrixAutoUpdate === true ) {
		
		this.updateMatrix();
		
	} else {
		
		this._matrixWorldNeedsUpdate = true;
		
	} 

	var parent = this.parent; 

	if ( this._matrixWorldNeedsUpdate === true ) {
			
		if ( parent === undefined ) {

				this._matrixWorld.copy( this._matrix );				
			
		} else {
			
			this._matrixWorld.multiplyMatrices( parent._matrixWorld, this._matrix );

		}

	}
	
	this._matrixWorldNeedsUpdate = false;
	
	return this._matrixWorld;
	
},

updateAncestorsMatrixWorld: function () {

	var parent = this.parent; 

	if ( parent !== undefined ) {

		parent.updateAncestorsMatrixWorld();

	}

	this.updateMatrixWorld();

	return this._matrixWorld;

},

updateChildrenMatrixWorld: function () {

	this.updateMatrixWorld();

	var children = this.children;

	for ( var i = 0, l = children.length; i < l; i ++ ) {

		children[ i ].updateChildrenMatrixWorld();

	}

	return this._matrixWorld;

},
```

Instead of having just a single method that can traverse up or down, we have three methods, one that only updates the matrixWorld, one that traverses up and one that traverses down.

Would this work?

WestLangley commented:
This may be related to @mrdoob's concern...

With the proposed API to `updateMatrixWorld()`, the 2nd arg is really private. Parents must always be checked to guarantee the correct answer -- unless it is known in advance that the parents are already correct.

So `updateParents` should only be set to `false` internally, to prevent bubbling up when traversing down the scene graph. It is an arg that the user should not set.

    updateMatrixWorld( updateChildren, _updateParents )

So we have an exposed arg that should not really be exposed.

gero3 commented:
I'm actually thinking of removing updateChildren completely. Because right now in this branch I'm actually calculating the matrices twice in the renderer and this doesn't affect the performance as bad as I thought it would. To make it possible to keep a scene static, I would even create a autoupdateMatrixworld property to make it  possible to not even do any calculations to calculate the Matrices.

WestLangley commented:
> I would even create a autoupdateMatrixworld property

Internally, that would be fine. As I said earlier, from and API standpoint, I think an `Object3D.static` property is better. It would apply to the object and all its children.

gero3 commented:
Let's leave static or autoupdateMatrixworld out of this right now and create another PR for that.
Because that seems to me is another big discussion.

Removing updateChildren & updateParents completely or another implementation seems to me is the more pressing matter.

WestLangley commented:
@gero3 Fair enough.

mrdoob commented:
@gero3 my problem with the current implementation is that `updateMatrixWorld` is hard to understand the flow. Meaning that it goes up and down and you don't know where the code is being executed. I would prefer to have something more explicit like in my last comment. Do you see any issues with my proposed code?

gero3 commented:
The issue is updateChildrenMatrixWorld. This function should always be preceded with updateAncestorsMatrixWorld, I think to avoid all issues with matrices. Because It should always have an updated matrixWorld of the parent to continue working with. We may be able to work around this by documenting that this should always be used like this.

On a different note, if we use the proposed code, then we should rename them to the following:

- updateMatrixWorld -> _updateMatrixWorld
- updateAncestorsMatrixWorld -> updateMatrixWorld
- updateChildrenMatrixWorld -> updateChildrenMatrixWorld

The reason why is because they never should call _updateMatrixWorld outside of objects and updateMatrixWorld is the correct naming to update the matrix world correctly.

I understand the reason why you think it is confusing, that's why I suggested to leave updating children out of it. Since it seem to have little incentive to keep it around.

WestLangley commented:
I agree with everything @gero3 said. And, as I said earlier, my request would be for `object.updateMatrixWorld()` to do just what its name says it does -- correctly update the world matrix of the object only -- not the children.

> I suggested to leave updating children out of it. Since it seem to have little incentive to keep it around.

@gero3 Have you demonstrated that there is no longer a compelling reason for the renderer to traverse the scene and update the world matrices?

gero3 commented:
I'll generate an example for this.

mrdoob commented:
> I'll generate an example for this.

Thanks!

dubejf commented:
> As I said earlier, from and API standpoint, I think an `Object3D.static` property is better.

You'll need a synonym. `static` is a reserved keyword in JS.

mrdoob commented:
Btw...

> The issue is updateChildrenMatrixWorld. This function should always be preceded with updateAncestorsMatrixWorld, I think to avoid all issues with matrices.

Yep. That wouldn't be a problem. In fact, it would help the user understand what's happening.

gero3 committed:
remove updatematrixWorld in the renderers

gero3 commented:
[my version](https://preview.c9.io/gero3/three_js/examples/webgl_performance_deep.html?_c9_id=livepreview1&_c9_host=https://ide.c9.io)
[dev version](https://preview.c9.io/gero3/dev_three_js/three.js/examples/webgl_performance_deep.html?_c9_id=livepreview0&_c9_host=https://ide.c9.io)

Sorry for the bad example but I couldn't put a lot of time in it.

It is a scene with 10 object deep and then all meshes stored in the deepest object. And the lowest object3d is rotated.

WestLangley commented:
Depending on the position of the cursor, I am getting a frame rate of 27 - 40 with your example; 40 - 60 with the dev version.

BTW, the far plane is clipping, which distorts the results, I expect.

mrdoob commented:
`dev version` is ~57fps, `my version` is ~34fps 😕

gero3 commented:
@mrdoob, @westlangley, What do you think about the following??
I don't mind have more function as long as only 1 needs to be updated for inherintance.

	_updateMatrixWorld: function () {

		if ( this.matrixAutoUpdate === true ) {

			this.updateMatrix();

		} else {

			this._matrixWorldNeedsUpdate = true;

		} 

		var parent = this.parent; 

		if ( this._matrixWorldNeedsUpdate === true ) {

			if ( parent === undefined ) {

				this._matrixWorld.copy( this._matrix );             

			} else {

				this._matrixWorld.multiplyMatrices( parent._matrixWorld, this._matrix );

			}

		}

		this._matrixWorldNeedsUpdate = false;

		return this._matrixWorld;

	},

	updateMatrixWorld: function () {

		var parent = this.parent; 

		if ( parent !== undefined ) {

			parent.updateAncestorsMatrixWorld();

		}

		this.updateMatrixWorld();

		return this._matrixWorld;

	},

	_updateChildrenMatrixWorld: function () {

		this._updateMatrixWorld();

		var children = this.children;

		for ( var i = 0, l = children.length; i < l; i ++ ) {

			children[ i ]._updateChildrenMatrixWorld();

		}

		return this._matrixWorld;

	},

	updateChildrenMatrixWorld: function () {

		if ( parent !== undefined ) {
			this.parent.updateMatrixWorld();
		}
		this._updateChildrenMatrixWorld();

		return this._matrixWorld;

	},

WestLangley commented:
I think your version of `updateMatrixWorld()` should call `_updateMatrixWorld()`, instead of itself.

Also, I would prefer something like this, instead:


	_updateMatrixWorld: function () {

		if ( this.matrixAutoUpdate === true ) {

			this.updateMatrix();

		} else {

			this._matrixWorldNeedsUpdate = true;

		} 

		var parent = this.parent; 

		if ( this._matrixWorldNeedsUpdate === true ) {

			if ( parent === undefined ) {

				this._matrixWorld.copy( this._matrix );             

			} else {

				this._matrixWorld.multiplyMatrices( parent._matrixWorld, this._matrix );

			}

		}

		this._matrixWorldNeedsUpdate = false;

		return this._matrixWorld;

	},

	_updateChildrenMatrixWorld: function () {

		this._updateMatrixWorld();

		var children = this.children;

		for ( var i = 0, l = children.length; i < l; i ++ ) {

			children[ i ]._updateChildrenMatrixWorld();

		}

		return this._matrixWorld;

	},

	updateMatrixWorld: function ( updateChildren ) {

		var parent = this.parent; 

		if ( parent !== undefined ) {

			parent.updateMatrixWorld();

		}

		if ( updateChildren === true ) {

			this._updateChildrenMatrixWorld();

		} else {

			this._updateMatrixWorld();

		}

		return this._matrixWorld;

	},

gero3 commented:
I'm unbiased between both implementations actually. And you were right about calling _updateMatrixWorld in updateMatrixWorld . I've updated my code. btw can you use  ```javascript and ``` around the code. It makes the code highlight.

WestLangley commented:
>  btw can you use javascript and around the code. It makes the code highlight.

I indent the code with 4 spaces or a tab. It highlights for me. Not for you?

WestLangley commented:
Wait. I see the difference, now.

WestLangley commented:
I prefer the single, exposed method `updateMatrixWorld()`.

@gero3 1. Do we agree that users should not have to call this method? 2. Do we agree that the renderer should not have to call it, either? 3. If the renderer does call it, is there a benefit?

gero3 commented:
The reason we need a function that updates the children is becuase we can have a lot of children with more then 1 ancestor. Which would mean if we kept calling matrixWorld. It would keep updating the ancestors for every child. This starts to give troubles as shown with my examples before.

> 1. Do we agree that users should not have to call this method?
Users should never call this method. 

> 2. Do we agree that the renderer should not have to call it, either?
I would call this method in renderer becuase of performance issues otherwise with deep objects.

 > 3. If the renderer does call it, is there a benefit?
performance in deep scenes which is what my examples show.

WestLangley commented:
@gero3 Excellent answers. This is really fine work, IMHO. Thank you.

gero3 commented:
@mrdoob 

What do you think about these??

- https://github.com/mrdoob/three.js/pull/6851#issuecomment-127148195
- https://github.com/mrdoob/three.js/pull/6851#issuecomment-127142344

I can make a new PR with this solution then.

mrdoob commented:
Looking good!

However, for code clarity reasons, I still think we can afford having more explicit methods.
Even if it's more verbose, I think it's easier to follow this:

```javascript
this.updateAncestorsMatrixWorld();
this.updateMatrixWorld();
this.updateChildrenMatrixWorld();
```

Instead of this:

```javascript
this.updateMatrixWorld( true );
```

gero3 commented:
succeeded by #7019
